readme.md
## oc Test middleware service

This is a middleware server for testing owncloud (owncloud10 or OCIS) with different Clients. This server handles the operations which are required to "get the server to a certain state"

### How it works
Since most tests in owncloud are written in gherkin, the middleware server accepts a gherkin step, parses it and runs the appropriate code to get the server to a certain state.

For example, If we want to run a gherkin step `Given user "Alice" has been created with default attributes`, the test runner sends this step to the middleware server which runs it and creates the user on the oc server with all necessary default attributes (displayname, email etc.) and also stores the user so that we can cleanup after the test is complete.

### Endpoints
- POST /init
    
    reset the middleware state and start a test scenario execution
    ```
    curl -XPOST http://localhost:3000/init
    ```

- POST /execute

    execute a gherkin step
    ```
    curl -XPOST http://localhost:3000/execute -d '{"step": "Given user \"Alice\" has been created with default attributes and without skeleton files"}' -H "Content-Type: application/json"
    ```

    Gherkin data tables are passed as a simple 2D array in json
    ```
    curl -XPOST http://localhost:3000/execute -d '{"step":"Given these groups have been created:","table":[["groupname"],["HelloGroup"]]}' -H 'Content-Type: application/json'
    ```

- GET /state

    retrieve current state of the test runner.
    information like created users/groups list are shared
    ```shell script
      curl http://localhost:3000/state
    ```


- POST /cleanup

    cleanup the middleware state and the owncloud server
    ```
    curl -XPOST http://localhost:3000/cleanup
    ```

### Integration with test frameworks
The middleware service is designed to be used with gherkin test runners such as cucumber and behat. In order to integrate this service with the test suite, we need to find a way to capture the steps that need to be run in the middleware. Most test runners allow use of regex in the matching of the step definition. We can use any regex rule to match all the steps that need to run in the middleware. you can capture them at once and send them to the middleware service.
for example, for all the steps you want to run in the middleware end them with `in the server`, then you can use regex `/^(.*) in the server$/` to capture all the steps

eg. Cucumber Integration

```gherkin
# test.feature
Feature: test feature

  Background:
    Given user "Alice" has been created with default attributes in the server
    And user "Alice" has created folder "new folder" in the server
  
  Scenario: ...
```

```js
// someContext.js
function handler(statement) {
    return fetch("http://localhost:3000/execute", {
        method: "POST",
        body: JSON.stringify({step: "Given " + statement}),
        headers: {
            'Content-Type': 'application/json'
        },
    }).then(res => {
        if (res.ok) {
            // somewhere in the middle of init and cleanup
            if(currentStateIsToBeFetched) {
              return fetch("http://localhost:3000/state").then(res => {
                return res.json()
              })
            }
            return res.text()
        } else {
            throw new Error(res.text())
        }
    }).catch(err => {
        console.error(err)
        return Promise.reject(err)
    })
}

Before(function() {
    return fetch("http://localhost:3000/init", {
        method: "POST"
    })
})

After(function() {
    return fetch("http://localhost:3000/cleanup", {
        method: "POST"
    })
})

Given(/^(.*) in the server$/, handler);
```

**notes**
- You will need to create a separate matcher if you want to send data tables because cucumber will not match the same step with an optional datatable.
- Since most of the test runner state such as the list of users created by the test runner are stored in the middleware, there is no need to remember them in the test runner itself.

### Config
#### Server requirements:
- `testing` app must be enabled to use steps that run `occ` commands or set ownCloud 10 system settings (see [testing/README.md](https://github.com/owncloud/testing/blob/master/README.md))

Note: the `testing` app is not available on oCIS. When such steps are used against an oCIS server they will not be effective.

#### Server specific config variables

| setting | meaning | default |
|-|-|-|
| `HOST` | Host for the middleware server | localhost |
| `PORT` | Port for the middleware server | 3000 |

#### Test Specific config variables

| setting | meaning | default |
|-|-|-|
| `BACKEND_HOST`      | ownCloud server URL (or reva service url for running with OCIS) | http://localhost:8080 |
| `BACKEND_USERNAME` | ownCloud administrator username | admin                 |
| `BACKEND_PASSWORD` | ownCloud administrator password  | admin                 |
| `REMOTE_BACKEND_HOST` | ownCloud remote server URL | http://localhost:8080 |
| `RUN_ON_OCIS` | Running the tests using the OCIS backend | false |
| `OCIS_REVA_DATA_ROOT` | Data directory of OCIS  | /var/tmp/reva |
| `OCIS_SKELETON_DIR` | Skeleton files directory for new users | - |

#### Ldap Specific config variables
| setting | meaning | default |
| -- | -- | -- |
| `RUN_WITH_LDAP` | use LDAP user backend | false |
| `LDAP_SERVER_URL`  | Url of the ldap server | ldap://127.0.0.1 |
| `LDAP_ADMIN_PASSWORD`  | admin password of the ldap server | admin |
| `LDAP_BASE_DN` | base DN of the admin server | cn=admin,dc=owncloud,dc=com |

#### Graph Specific config variables
| setting | meaning | default |
| -- | -- | -- |
| `TEST_WITH_GRAPH_API` | use graph api to create users | false |

### Starting the server
To start the middleware server change into the checkout directory of owncloud-test-middleware and use the following commands
```
yarn install  # install dependencies from package.json
yarn start
```
This command assumes that your backend server is running on `http://localhost:8080/` for oc10 (or `https://localhost:9200` when `RUN_WITH_OCIS` is set). If your backend is running on a different address use:
```
BACKEND_HOST=http://localhost/owncloud-server yarn start
```


CHANGELOG.md
## Changelog for [1.8.8] (2023-12-06)
The following sections list the changes for 1.8.8

[1.8.8]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.7...v1.8.8

### Summary
- Make webdavContext.getResourceType more resilient @dschmidt in https://github.com/owncloud/owncloud-test-middleware/pull/142

## Changelog for [1.8.7] (2023-11-16)
The following sections list the changes for 1.8.7

[1.8.7]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.6...v1.8.7

### Summary
- Update the status code expectation for create group @SwikritiT in https://github.com/owncloud/owncloud-test-middleware/pull/140

## Changelog for [1.8.6] (2023-11-15)
The following sections list the changes for 1.8.6

[1.8.6]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.5...v1.8.6

### Summary
- fix response code by @ScharfViktor in https://github.com/owncloud/owncloud-test-middleware/pull/137
- update docker plugin by @ScharfViktor https://github.com/owncloud/owncloud-test-middleware/pull/138

## Changelog for [1.8.5] (2023-03-02)
The following sections list the changes for 1.8.5

[1.8.5]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.4...v1.8.5

### Summary
- multiple retries webdav download on 425 status code by @kulmann in https://github.com/owncloud/owncloud-test-middleware/pull/134

## Changelog for [1.8.4] (2023-03-02)
The following sections list the changes for 1.8.4

[1.8.4]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.3...v1.8.4

### Summary
- retry webdav download on 425 status code by @kulmann in https://github.com/owncloud/owncloud-test-middleware/pull/132

## Changelog for [1.8.3] (2022-12-29)
The following sections list the changes for 1.8.3

[1.8.3]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.2...v1.8.3

### Summary
- Strip time from expiration date by @saw-jan in https://github.com/owncloud/owncloud-test-middleware/pull/129
- Pin node to version v16.3 and eslint parser to ecma2021 by @saw-jan in https://github.com/owncloud/owncloud-test-middleware/pull/128

## Changelog for [1.8.2] (2022-10-17)
The following sections list the changes for 1.8.2

[1.8.2]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.1...v1.8.2

### Summary
- [lint] set indentation rule by @kiranparajuli589 in https://github.com/owncloud/owncloud-test-middleware/pull/125
- [tests-only]Refactor occ related step for ocis by @amrita-shrestha in https://github.com/owncloud/owncloud-test-middleware/pull/124

## Changelog for [1.8.1] (2022-08-05)
The following sections list the changes for 1.8.1

[1.8.1]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.8.0...v1.8.1

### Summary
- Fix user cleanup after tests finish @amrita-shrestha  [#121](https://github.com/owncloud/owncloud-test-middleware/pull/121)

## Changelog for [1.8.0] (2022-08-01)
The following sections list the changes for 1.8.0

[1.8.0]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.7.0...v1.8.0

### Summary
- Add created group to the array by @SwikritiT in [#118](https://github.com/owncloud/owncloud-test-middleware/pull/118)

## Changelog for [1.7.0] (2022-05-31)
The following sections list the changes for 1.7.0

[1.7.0]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.6.0...v1.7.0

### Summary
- [tests-only]Test for depth 1 instead of instead of infinty by @SagarGi in [#117](https://github.com/owncloud/owncloud-test-middleware/pull/117)

## Changelog for [1.6.0] (2022-4-29)
The following sections list the changes for 1.6.0

[1.6.0]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.5.0...v1.6.0

### Summary
- Creating group and adding member to group via GraphApi by @ScharfViktor in https://github.com/owncloud/owncloud-test-middleware/pull/113

## Changelog for [1.5.0] (2022-4-27)
The following sections list the changes for 1.5.0

[1.5.0]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.4.0...v1.5.0

### Summary
- Creating a user via graph Api by @ScharfViktor in https://github.com/owncloud/owncloud-test-middleware/pull/107
- Add content to file in Folder,with,coma by @pascalwengerter in https://github.com/owncloud/owncloud-test-middleware/pull/109
- Fix lint and lint packages by @kiranparajuli589 in https://github.com/owncloud/owncloud-test-middleware/pull/106
- [enhancement] Adjust express endpoints by @kiranparajuli589 in https://github.com/owncloud/owncloud-test-middleware/pull/108
- encode uri for the provisioning urls by @kiranparajuli589 in https://github.com/owncloud/owncloud-test-middleware/pull/105

## Changelog for [1.4.0] (2022-02-16)

The following sections list the changes for 1.4.0

[1.4.0]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.3.0...v1.4.0

### Summary
- Fix local file create step by @kiranparajuli589 in https://github.com/owncloud/owncloud-test-middleware/pull/98
- Fixed duplicated imports by @kiranparajuli589 in https://github.com/owncloud/owncloud-test-middleware/pull/99
- added steps to check items count inside a folder by @sushmita56 in https://github.com/owncloud/owncloud-test-middleware/pull/100
- Fix typo of 'actual' by @phil-davis in https://github.com/owncloud/owncloud-test-middleware/pull/102

## Changelog for [1.3.0] (2022-01-25)

The following sections list the changes for 1.2.0

[1.3.0]: https://github.com/owncloud/owncloud-test-middleware/compare/v1.2.0...v1.3.0

### Summary
- Used entrypoint to start the middleware (docker) [#93](https://github.com/owncloud/owncloud-test-middleware/pull/93)
- Added given step to create folder with files by [#92](https://github.com/owncloud/owncloud-test-middleware/pull/92)
- Retry download if resource is locked [#90](https://github.com/owncloud/owncloud-test-middleware/pull/90)
- Tidy some code [#89](https://github.com/owncloud/owncloud-test-middleware/pull/89)

## Changelog for [1.2.0] (2022-01-18)

The following sections list the changes for 1.2.0


src/config.js
const path = require('path')

const withHttp = (url) => (/^https?:\/\//i.test(url) ? url : `http://${url}`)

const RUN_WITH_LDAP = !!process.env.RUN_WITH_LDAP
const RUN_ON_OCIS = process.env.RUN_ON_OCIS === 'true'
const TEST_WITH_GRAPH_API = process.env.TEST_WITH_GRAPH_API === 'true'

const LOCAL_BACKEND_URL = withHttp(
  process.env.BACKEND_HOST || (RUN_ON_OCIS ? 'https://localhost:9200' : 'http://localhost:8080')
)
const REMOTE_BACKEND_URL = process.env.REMOTE_BACKEND_HOST
  ? withHttp(process.env.REMOTE_BACKEND_HOST || 'http://localhost:8080')
  : undefined
const BACKEND_ADMIN_USERNAME = process.env.BACKEND_USERNAME || 'admin'
const BACKEND_ADMIN_PASSWORD = process.env.BACKEND_PASSWORD || 'admin'

const REMOTE_UPLOAD_DIR = process.env.REMOTE_UPLOAD_DIR || path.join(__dirname, '../filesForUpload')

const OCIS_REVA_DATA_ROOT = process.env.OCIS_REVA_DATA_ROOT || '/var/tmp/ocis/storage/owncloud'
const LDAP_SERVER_URL = process.env.LDAP_SERVER_URL || 'ldap://127.0.0.1'
const LDAP_BASE_DN = process.env.LDAP_BASE_DN || 'cn=admin,dc=owncloud,dc=com'
const LDAP_ADMIN_PASSWORD = process.env.LDAP_ADMIN_PASSWORD || 'admin'
const TESTING_DATA_DIR = process.env.TESTING_DATA_DIR || path.join(__dirname, '../data')

const assert = require('assert')

if (RUN_ON_OCIS) {
  process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
}

const config = {
  globals: {
    filesForUpload: REMOTE_UPLOAD_DIR,
    backend_url: LOCAL_BACKEND_URL,
    remote_backend_url: REMOTE_BACKEND_URL,
    backend_admin_username: BACKEND_ADMIN_USERNAME,
    backend_admin_password: BACKEND_ADMIN_PASSWORD,
    default_backend: 'LOCAL',
    ocis: RUN_ON_OCIS,
    ldap: RUN_WITH_LDAP,
    ldap_url: LDAP_SERVER_URL,
    ocis_data_dir: OCIS_REVA_DATA_ROOT,
    ldap_base_dn: LDAP_BASE_DN,
    testing_data_dir: TESTING_DATA_DIR,
    ldap_password: LDAP_ADMIN_PASSWORD,
    graph: TEST_WITH_GRAPH_API,
  },
  assert,
}

module.exports.client = config
